### ISSUE TITLE
Visual Studio 2019 Support

### ISSUE THREAD

**Issue by devSoheilAlizadeh** [RazorGenerator/RazorGenerator#188]
Razor Generator Extension doesn't support visual studio 2019

**Comment by StephanBis** [comment]
Any update on this?

**Comment by StephanBis** [comment]
All checks seem to have passed, can you merge the PR? Thank you!

**Comment by Jehoel** [comment]
@devSoheilAlizadeh I've built it for VS2019 using the latest preview release of VS2019 but the VSIX installer doesn't recognize my VS2019 install. I'm investigating.

**Comment by Daddoon** [comment]
@devSoheilAlizadeh Jehoel resolved the issue on #189  .

**Comment by Jehoel** [comment]
I'm leaving this issue open until I'm able to actually publish the new VSIX to the Marketplace. I've reached out @davidebbo and @pranavkm by e-mail to see if they can add me to the Publisher account.

**Comment by Jehoel** [comment]
I now have RazorGenerator's Custom Tool working in VS2019 - phew. The VSIX is also compatible with VS2015 and VS2017.

Caveats:

The code in RazorGenerator to load the "v1" (ASP.NET 3.5, .NET Framework 2.0-3.5), and "v2" (ASP.NET 4.0) versions of the Razor libraries are... convoluted. I think it would be prudent to simply cut-off those vestigal versions now, no?
No support for ASP.NET Core's version of Razor. This is increasingly more important so I'll see what I can do.

-------

I'll clean up the duplicate issues relating to VS2017/VS2019 support soon.

**Comment by Daddoon** [comment]
> 
> 
> I now have RazorGenerator's Custom Tool working in VS2019 - phew. The VSIX is also compatible with VS2015 and VS2017.
> 
> Caveats:
> 
> The code in RazorGenerator to load the "v1" (ASP.NET 3.5, .NET Framework 2.0-3.5), and "v2" (ASP.NET 4.0) versions of the Razor libraries are... convoluted. I think it would be prudent to simply cut-off those vestigal versions now, no?
> No support for ASP.NET Core's version of Razor. This is increasingly more important so I'll see what I can do.
> 
> I'll clean up the duplicate issues relating to VS2017/VS2019 support soon.

Is this also related to the fact that since some days, I have issue with intellisense doing things completly wrong if i try to do an autocompletion ? Like the autocomplete does not appear where it should, and instead write code somewhere else in the file.

**Comment by davidebbo** [comment]
> The code in RazorGenerator to load the "v1" (ASP.NET 3.5, .NET Framework 2.0-3.5), and "v2" (ASP.NET 4.0) versions of the Razor libraries are... convoluted. I think it would be prudent to simply cut-off those vestigal versions now, no?

Yes, limiting the support matrix seems like the right thing to do. People need to move forward!

**Comment by Jehoel** [comment]
@davidebbo I need to pick your brain for some guidance: when RazorGenerator (and the Razor assemblies) are ran as a **Custom Tool** in VS2019 they're hosted in-proc (though in their own appdomain) inside a  .NET Framework 4.7.2+ appdomain - from what I can tell the v1 and v2 assemblies worked this way by sheer miracle of .NET back/forwards compatibility. I understand this approach simply won't work at all for .NET Core Razor assemblies because .NET Core doesn't support AppDomains and VS2019 don't run .NET Core code in-proc for the same reason (well, and other sensible reasons too) - so I'm thinking of supporting the .NET Core version as a Custom Tool by simply invoking the MSBuild version out-of-proc - the downside is losing integration with EnvDTE etc from within RazorGenerator-managed `.cshtml` files - is that the right approach, or is there an approved way to run .NET Core-based Custom Tool assemblies?

**Comment by Jehoel** [comment]
@Daddoon 

> Is this also related to the fact that since some days, I have issue with intellisense doing things completly wrong if i try to do an autocompletion ? Like the autocomplete does not appear where it should, and instead write code somewhere else in the file.

RazorGenerator builds on-top of VS's existing Razor tooling - with some creative workarounds for scenarios unsupported by VS (like how originally if you had a non-web project VS wouldn't recognize `.cshtml` files at all). I've forgotten how all of this works (this is _embarrassing_ given I used to work on Visual Studio when I was at MS, albeit not in the main IDE Text Editor team) - but the exact issue you're describing sounds like something I'm experiencing myself when editing `.aspx` files (don't ask) especially when VS hasn't fully _warmed up_ - it's something I started noticing around August 2020 when (_I think_) they reworked the HTML editing experience (shared by Razor `.cshtml`, HTML, and WebForms `.aspx`) to accomodate changes to Roslyn. I'm not an expert here and I don't personally know anyone still on the editor team who can lend any insight.

I'm sure that eventually I'll have to confront this issue for RazorGenerator but I haven't gotten to that point yet so I don't have anything helpful to say yet.

**Comment by davidebbo** [comment]
@Jehoel .NET Core support seems like a big topic which should be discussed in a separate issue, since this one is tracking VS 2019 support.

I think we'll need to add big questions first. Like whether there is even demand for this tool in the Core world.

Frankly, I'm very disconnected from the .NET world since leaving MS in 2018, so I'm not at all up to date on the latest and greatest things happening with .NET core. Maybe we can get some thoughts from Core experts (but let's create a separate issue).

@pranavkm are you still connected to the team?

**Comment by pranavkm** [comment]
👋🏽 @davidebbo 

* The Razor language assemblies target netstandard2.0, so they should be loadable inside VS 2019 which requires .NET 4.8.
* .NET Core ships a build-time experience for Razor files and it's on by default. Things like build time errors and faster startup that RazorGenerator tries to solve are kinda taken care of. 
* The templating aspect is still unsolved i.e. trying to use .cshtml files as standalone tools, but the runtime (.NET Fx vs .NET Core) should be kinda moot there - you're not really using runtime features in that case.

**Comment by Jehoel** [comment]
@pranavkm Phew, that's a relief - I was concerned that the latest Razor-for-.NET 5 stuff wouldn't be targeting .NET Standard anymore.
